cmd/access/node_builder/access_node_builder.go

```diff
@@ -590,14 +626,120 @@ func (builder *FlowAccessNodeBuilder) BuildExecutionSyncComponents() *FlowAccess
 				builder.State,
 				builder.Storage.Headers,
 				builder.executionDataConfig,
+				execDataDistributor,
 			)
 
 			builder.FollowerDistributor.AddOnBlockFinalizedConsumer(builder.ExecutionDataRequester.OnBlockFinalized)
-			builder.ExecutionDataRequester.AddOnExecutionDataReceivedConsumer(execDataDistributor.OnExecutionDataReceived)
+
+			// add requester into ReadyDoneAware dependency passed to indexer. This allows the indexer
+			// to wait for the requester to be ready before starting.
+			requesterDependable.Init(builder.ExecutionDataRequester)
 
 			return builder.ExecutionDataRequester, nil
 		})
 
+	if builder.executionDataIndexingEnabled {
+		var indexedBlockHeight storage.ConsumerProgress
+
+		builder.
+			Module("indexed block height consumer progress", func(node *cmd.NodeConfig) error {
+				// Note: progress is stored in the MAIN db since that is where indexed execution data is stored.
+				indexedBlockHeight = bstorage.NewConsumerProgress(builder.DB, module.ConsumeProgressExecutionDataIndexerBlockHeight)
+				return nil
+			}).
+			Module("events storage", func(node *cmd.NodeConfig) error {
+				builder.Storage.Events = bstorage.NewEvents(node.Metrics.Cache, node.DB)
+				return nil
+			}).
+			Module("transaction results storage", func(node *cmd.NodeConfig) error {
+				builder.Storage.LightTransactionResults = bstorage.NewLightTransactionResults(node.Metrics.Cache, node.DB, bstorage.DefaultCacheSize)
+				return nil
+			}).
+			DependableComponent("execution data indexer", func(node *cmd.NodeConfig) (module.ReadyDoneAware, error) {
+				// Note: using a DependableComponent here to ensure that the indexer does not block
+				// other components from starting while bootstrapping the register db since it may
+				// take hours to complete.
+
+				pdb, err := pStorage.OpenRegisterPebbleDB(builder.registersDBPath)
+				if err != nil {
+					return nil, fmt.Errorf("could not open registers db: %w", err)
+				}
+				builder.ShutdownFunc(func() error {
+					return pdb.Close()
+				})
+
+				bootstrapped, err := pStorage.IsBootstrapped(pdb)
+				if err != nil {
+					return nil, fmt.Errorf("could not check if registers db is bootstrapped: %w", err)
+				}
+
+				if !bootstrapped {
+					// since indexing is done using execution data, make sure that execution data
+					// will be available starting at the checkpoint height. otherwise, the indexer
+					// will fail to progress.
+					if builder.checkpointHeight < builder.executionDataConfig.InitialBlockHeight {
```

Comment:
This is out of the scope of this PR but is there no way to get the height of the checkpoint from the checkpoint? This feels like we could set it to the wrong value or skip some heights by mistake. 
Also, is the check sufficient? What if the checkpoint height is bigger than the initial height of exe data, isn't that unexpected too? Shouldn't the height always be -1 the exe data height? What is the case where this is not true?

Reply:
This check is sufficient for root block, but is not a check we should keep if we're using a checkpoint generated by dynamic bootstrap later, but we're at least 3 months away from that

Reply:
> is there no way to get the height of the checkpoint from the checkpoint?

my understanding is that checkpoints currently do not include the height. That is coming in a future improvement from the storehouse work.

> Also, is the check sufficient? What if the checkpoint height is bigger than the initial height of exe data, isn't that unexpected too? Shouldn't the height always be -1 the exe data height? What is the case where this is not true?

At the beginning of the spork, `checkpointHeight` should the the same as `InitialBlockHeight`. Both should be the root height. For nodes starting after the spork, or using a smaller execution data retention window, they may not be the same.

My 2 cents is it's better to be flexible where possible (and safe). that way the software can handle situations we may not anticipated, like recovering from an outage. In this case, we just need to make sure that the configuration allows for indexing to work. As long as there is execution data available from `checkpoint + 1` and higher, the indexer will be able to progress.

Reply:
Good thinking.